### PR TITLE
Added support for $ref in allOf

### PR DESCRIPTION
**Pull request by prevuelta**
#207

### CODE REVIEW

**Comment by ErikWittern** [comment]
@prevuelta Thank you very much for taking this one on! It seems to me that by changing the `example_oas.json` to use a `$ref` within an `allOf` definition, existing test cases now also cover this behavior, correct?

**Comment by Alan-Cha** [comment]
Thanks for the PR! Correct me if I’m wrong but I believe this strategy can only handle top-level references. If references appear within a schema, things will still break. This is still a good first step in the right direction.

**Comment by Alan-Cha** [comment]
I think what we need to do is recurse early, first creating data definitions for the subschemas before finishing the `allOf` data definition. We will also need to make sure that we don’t enter any infinite loops as it is possible. This is all assuming that we continue with trying to resolve references and JSON combine schema mechanisms ourselves... which I think is still up in the air, at least until we get feedback regarding this [issue](https://github.com/APIDevTools/json-schema-ref-parser/issues/70#issuecomment-511540802).

**Comment by prevuelta** [comment]
I have expanded this in my fork to recursively deference and merge allOf - also implemented for array items. Should I create another pr or leave this for now?

**Comment by Alan-Cha** [comment]
@prevuelta If you could create another PR, that would be nice! 😄

**Comment by Alan-Cha** [comment]
It took a while to find time to look into this but now we have not only support for references in `allOf` but also support for nested `allOfs`.

File: packages/openapi-to-graphql/src/preprocessor.ts [hunk 1]
```diff
@@ -733,3 +695,63 @@ function getSchemaName(
 
   return schemaName
 }
+
+/**
+ * Recursively add the (nested) allOf schemas to root-level data defintion
+ *
+ * @param def Root-level data definition
+ */
+function addAllOfToDataDef(
```
[the remark this answers is not in the document]
Comment (ErikWittern) [reply packages/openapi-to-graphql/src/preprocessor.ts]: Looks clean.

File: packages/openapi-to-graphql/src/utils.ts [hunk 2]
```diff
@@ -20,7 +20,6 @@ export const mitigations = {
   MISSING_RESPONSE_SCHEMA: `Ignore operation.`,
   DUPLICATE_FIELD_NAME: `Ignore field and maintain preexisting field.`,
   DUPLICATE_LINK_KEY: `Ignore link and maintain preexisting link.`,
-  UNRESOLVABLE_REFERENCE: `The schema will not be resolved.`,
```
[the remark this answers is not in the document]
Comment (ErikWittern) [reply packages/openapi-to-graphql/src/utils.ts]: I think it is good, for now, to throw an error. If these cases become frequent, we should have a dedicated discussion on possible mitigations. Using graphql-json-type is definitely one option.

**Comment by ErikWittern** [comment]
@prevuelta @Alan-Cha Great work, thank you!

**Pull request merged**